daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/parsers/SequenceParserBases.scala

```diff
@@ -30,24 +36,58 @@ import org.apache.daffodil.processors.TermRuntimeData
 import org.apache.daffodil.processors.Failure
 import org.apache.daffodil.processors.ModelGroupRuntimeData
 
+abstract class UnorderedSequenceParserBase(
+  srd: SequenceRuntimeData,
+  choiceParser: ChoiceParser)
+  extends SequenceParserBase(srd, Vector(choiceParser)) {
+
+  override protected def parse(pstate: PState) = {
+
+    var isDone = false
+
+    while (!isDone && (pstate.processorStatus eq Success)) {
+      val priorPos = pstate.bitPos0b
+      choiceParser.parse(pstate)
+      val currentPos = pstate.bitPos0b
+
+      if ((pstate.processorStatus eq Success) && (currentPos <= priorPos)) {
+        isDone = true
+      }
+    }
+
+    pstate.setSuccess()
+    val erds = pstate.infoset.asInstanceOf[DIComplex].flattenAndValidateChildNodes()
+    pstate.infoset.asInstanceOf[DIComplex].sortChildNodesByPosition()
+
+    // We have encountered a min/maxOccurs validation error
+    if (erds.isDefined) {
+      erds.get.foreach(e => {
+        val min = e.minOccurs
+        val max = e.maxOccurs
+        val isUnbounded = max == -1
+
+        if (isUnbounded)
+          pstate.validationError("Element %s failed check of minOccurs='%s' and maxOccurs='UNBOUNDED", e.diagnosticDebugName, min)
+        else
+          pstate.validationError("Element %s failed check of minOccurs='%s' and maxOccurs='%s'", e.diagnosticDebugName, min, max)
+      })
```

Comment:
You are right, for some reason I thought I needed this logic outside of the infoset, but I don't think that is the case.  This should also simplify flattendAndValidate code too, instead of having to maintain a list to return.

Reply:
Actually, is there a way to get the PState object from the infoset? I don't see anything obvious, which I think might be why I did the actual validation error in SequenceParserBases

Reply:
I guess you coul pass the pstate into this function. Not sure if we do that anywhere is the infoset code, but it doesn't seem totally unreasonable.

Reply:
Actually, you could make the parameter a ThrowsSDE, and then pass the pstate into that. That way if other things needed to use that that had a ThrowsSDE context but not a PState it coudl still be used. Makes it less tied to parsing. I *think* ThrowsSDE also can also create validation errors?